# APS1: EDA

Realizando importações necessarias para o projeto

In [2]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.metrics import accuracy_score

## 1. Carregamento e inspeção inicial dos dados ##

Explicando cada feature e identificando os tipos de dados do dataset

- `age`:    **Idade do indivíduo** (contínua);
- `workclass`:  **Categoria do trabalho** (categórica);
- `fnlwgt`:     **Peso final** (contínua);
- `education`:  **Nível de escolaridade** (categórica);
- `education-num`:  **Nível de escolaridade em grau** (contínua);
- `marital-status`:     **Estado civil** (categórica);
- `occupation`:     **Ocupação profissional** (categórica);
- `relationship`:   **Relacionamento** (categórica);
- `race`:   **Etnia** (categórica);
- `sex`:    **Gênero** (categórica);
- `capital-gain`:   **Ganhos de capital** (contínua);
- `capital-loss`:   **Perdas de capital** (contínua);
- `hours-per-week`:     **Horas de trabalho semanais** (contínua);
- `native-country`:     **País de origem** (categórica);
- `income`:     **Renda anual** (contínua);

Carregamento e analise dos datasets:
- Conjunto de treino: `adult.data`

In [ ]:
# Carregando o dataset adult.data
df_train = pd.read_csv('dataset/adult.data', header=None)
df_train.columns = ['age', 'workclass', 'fnlwgt', 'education', 'education-num',
              'marital-status', 'occupation', 'relationship', 'race', 'sex',
              'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
              'income']
df_train.head()

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


- Conjunto de teste: `adult.test`

In [ ]:
# Carregando o dataset adult.test
df_test = pd.read_csv('dataset/adult.test', header=None, skiprows=1)
df_test.columns = ['age', 'workclass', 'fnlwgt', 'education', 'education-num',
              'marital-status', 'occupation', 'relationship', 'race', 'sex',
              'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
              'income']
# df_test.head()

Verificando o número de instâncias e features

In [5]:
df_train.shape, df_test.shape

((32561, 15), (16281, 15))

In [9]:
df_train.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'income'],
      dtype='str')

Dentro do dataset de treino possuimos **32561** instâncias e **14** features (e 1 target)

Separacao Treino/Teste

In [11]:
# selecao de features e target
features = ['age', 'workclass', 'fnlwgt', 'education', 'education-num',
              'marital-status', 'occupation', 'relationship', 'race', 'sex',
              'capital-gain', 'capital-loss', 'hours-per-week', 'native-country']

# Criação das matrize de features e vetor de target
X = df_train[features]
y = df_train['income']

# separacao dos dados em treino e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Inspecionando o conjunto de treino e teste

In [15]:
X_train

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country
5514,33,Local-gov,198183,Bachelors,13,Never-married,Prof-specialty,Not-in-family,White,Female,0,0,50,United-States
19777,36,Private,86459,Assoc-voc,11,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,1887,50,United-States
10781,58,Self-emp-not-inc,203039,9th,5,Separated,Craft-repair,Not-in-family,White,Male,0,0,40,United-States
32240,21,Private,180190,Assoc-voc,11,Married-civ-spouse,Farming-fishing,Husband,White,Male,0,0,46,United-States
9876,27,Private,279872,Some-college,10,Divorced,Other-service,Not-in-family,White,Male,0,0,40,United-States
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
29802,47,Private,359461,Bachelors,13,Married-civ-spouse,Craft-repair,Husband,White,Male,0,0,40,United-States
5390,31,Private,147215,12th,8,Divorced,Other-service,Unmarried,White,Female,0,0,21,United-States
860,18,Private,216284,11th,7,Never-married,Adm-clerical,Own-child,White,Female,0,0,20,United-States
15795,50,Self-emp-not-inc,54261,HS-grad,9,Married-civ-spouse,Farming-fishing,Husband,White,Male,0,0,84,United-States


In [16]:
y_train

5514       >50K
19777      >50K
10781     <=50K
32240     <=50K
9876      <=50K
          ...  
29802     <=50K
5390      <=50K
860       <=50K
15795     <=50K
23654     <=50K
Name: income, Length: 26048, dtype: str

Detecção de valores ausentes ('?') e inconsistênias.

In [18]:
# remover espaços em branco antes do '?'
df_train.replace(' ?', '?', inplace=True)
missing_counts = df_train.isin(['?']).sum()

print("=====================================")
print("Contagem de '?' por coluna no treino:")
print("=====================================")
print(missing_counts)

Contagem de '?' por coluna no treino:
age                  0
workclass         1836
fnlwgt               0
education            0
education-num        0
marital-status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital-gain         0
capital-loss         0
hours-per-week       0
native-country     583
income               0
dtype: int64


Dentro do nosso dataset de treino as features: `workclass`, `occupation` e `native_country` possuem dados faltantes.

Alem disso, as colunas `education` e `education-num` são redundantes, pois representam os mesmos dados.

In [ ]:
# categorical_features = ['workclass', 'occupation', 'native-country']
# categorical_transformer = Pipeline(steps=[
#     ('imputer', SimpleImputer(strategy='most_frequent')),  # moda
#     ('onehot', OneHotEncoder(sparse_output=False))  # converte para um vetor de zeros e uns
# ])